program/rust/src/rust_oracle.rs

```diff
@@ -23,4 +51,105 @@ pub fn update_version(
     _instruction_data: &[u8],
 ) -> OracleResult {
     panic!("Need to merge fix to pythd in order to implement this");
+    // Ok(SUCCESS)
+}
+
+
+/// initialize the first mapping account in a new linked-list of mapping accounts
+/// accounts[0] funding account           [signer writable]
+/// accounts[1] new mapping account       [signer writable]
+pub fn init_mapping(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    instruction_data: &[u8],
+) -> OracleResult {
+    let [_funding_account, fresh_mapping_account] = match accounts {
+        [x, y]
+            if valid_funding_account(x)
+                && valid_signable_account(program_id, y, size_of::<pc_map_table_t>())
+                && valid_fresh_account(y) =>
+        {
+            Ok([x, y])
+        }
+        _ => Err(ProgramError::InvalidArgument),
+    }?;
+
+    // Initialize by setting to zero again (just in case) and populating the account header
+    clear_account(fresh_mapping_account)?;
+
+    let hdr = load::<cmd_hdr_t>(instruction_data);
+    {
+        let mut mapping_account = load_account_as_mut::<pc_map_table_t>(fresh_mapping_account)?;
+        mapping_account.magic_ = PC_MAGIC;
+        mapping_account.ver_ = hdr.ver_;
+        mapping_account.type_ = PC_ACCTYPE_MAPPING;
+        mapping_account.size_ =
+            (size_of::<pc_map_table_t>() - size_of_val(&mapping_account.prod_)) as u32;
+    }
+
+    Ok(SUCCESS)
+}
+
+fn valid_funding_account(account: &AccountInfo) -> bool {
+    account.is_signer && account.is_writable
+}
+
+fn valid_signable_account(program_id: &Pubkey, account: &AccountInfo, minimum_size: usize) -> bool {
+    account.is_signer
+        && account.is_writable
+        && account.owner == program_id
+        && account.data_len() >= minimum_size
+        && Rent::default().is_exempt(account.lamports(), account.data_len())
+}
+
+/// Returns `true` if the `account` is fresh, i.e., its data can be overwritten.
+/// Use this check to prevent accidentally overwriting accounts whose data is already populated.
+fn valid_fresh_account(account: &AccountInfo) -> bool {
+    let pyth_acc = load_account_as::<pc_acc>(account);
+    match pyth_acc {
+        Ok(pyth_acc) => pyth_acc.magic_ == 0 && pyth_acc.ver_ == 0,
+        Err(_) => false,
+    }
+}
+
+/// Sets the data of account to all-zero
+fn clear_account(account: &AccountInfo) -> Result<(), ProgramError> {
+    let mut data = account
+        .try_borrow_mut_data()
+        .map_err(|_| ProgramError::InvalidArgument)?;
+    let length = data.len();
+    sol_memset(data.borrow_mut(), 0, length);
+    Ok(())
+}
+
+/// Interpret the bytes in `data` as a value of type `T`
+fn load<T: Pod>(data: &[u8]) -> &T {
+    from_bytes(&data[0..size_of::<T>()])
+}
+
+/// Interpret the bytes in `data` as a mutable value of type `T`
+#[allow(unused)]
+fn load_mut<T: Pod>(data: &mut [u8]) -> &mut T {
+    from_bytes_mut(&mut data[0..size_of::<T>()])
```

Comment:
Use` try_from_bytes_mut` instead of `from_bytes_mut`, the latter panics, which makes debugging harder

Reply:
In the load functions I think it can panic if data is too short, which can happen for instructions.

Reply:
Maybe use `data.get(0..size_of::<T>())` instead of `data[0..size_of::<T>()]`

Reply:
If you do not use try, it would panic if data is too short or too long. it needs to be an exact match

Reply:
Actually I think the way you solve this is by checking the size of instruction data in the dispatcher